### DISCUSSION TITLE
The state file does not correspond to the given data file when using custom feature extraction method

### DISCUSSION THREAD

**Discussion by michagast** [asreview/asreview#1048] [Troubleshooting (report errors in the "Issues" tracker)] [answered]
Hi everyone,

I am back again this time with a question about asreview lab. I have currently made my own extension for feature extraction for asreview. This can be found [here](https://github.com/michagast/asreview/blob/main/asreviewcontrib/models/enron_fe.py).
The relevant code is as follows:
```

from asreview.models.feature_extraction.base import BaseFeatureExtraction
import numpy as np

import pandas as pd                 #For data science purposes
import re                           #For performing regex
import torch                        #For running models with cuda
from transformers import AutoTokenizer, AutoModelForSequenceClassification, pipeline

class Enron(BaseFeatureExtraction):
    """Custom feature extraction

    Feature extraction that generates features based on sentiment values and named entity recogntion among other things.
    """

    name = "enron"
    label = "Enron feature extraction"

    def __init__(self, *args, **kwargs):

        super(Enron, self).__init__(*args, **kwargs)
    def transform(self, texts):
        model = AutoModelForSequenceClassification.from_pretrained("distilbert-base-uncased-finetuned-sst-2-english")

        tokenizernlp = AutoTokenizer.from_pretrained("distilbert-base-uncased-finetuned-sst-2-english")
        sentiment_analysis = pipeline("sentiment-analysis", model=model, tokenizer=tokenizernlp,
                                           max_length=512,
                                           truncation=True, device=0)
        result = np.empty([0])
        for text in texts:
            np.append(result, sentiment_analysis(text))
        return result
```
 The feature extraction method properly shows up within asreview lab and can be selected. The program then takes a while to apply the feature extraction to the dataset but after a while it gives the following error:
```
ERROR:root:Project enron - ValueError: The state file does not correspond to the given data file, please use another state file or dataset.
Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\asreview\__main__.py", line 122, in <module>
    main()
  File "C:AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\asreview\__main__.py", line 80, in main
    entry.load()().execute(sys.argv[2:])
  File "C:\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\asreview\entry_points\lab.py", line 112, in execute
    main_run_model(argv)
  File "C:\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\asreview\webapp\run_model.py", line 200, in main
    raise err
  File "C:\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\asreview\webapp\run_model.py", line 186, in main
    train_model(args.project_id, args.label_method)
  File "C:\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\asreview\webapp\run_model.py", line 117, in train_model
    reviewer = get_reviewer(dataset=data_fp, mode="minimal", **asr_kwargs)
  File "C:\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\asreview\review\factory.py", line 243, in get_reviewer
    reviewer = MinimalReview(as_data,
  File "C:\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\asreview\review\minimal.py", line 21, in __init__
    super(MinimalReview, self).__init__(*args, **kwargs)
  File "C:\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\asreview\review\base.py", line 209, in __init__
    raise ValueError("The state file does not correspond to the "
ValueError: The state file does not correspond to the given data file, please use another state file or dataset.
ERROR:root:Error on training.
```
Long story short, apparently the state file doesn't correspond to the given datafile. However this seems weird as this happens after I have selected the configuration and have asked asreview to train a base model with the given configurations and have thus not changed the dataset in any way. Note that it takes a while for this error to be generated as it seems that feature extraction is working but then suddenly it throws the error. After the error it is not possible to continue to the active learning.

Does anybody have a clue as to what causes this error? Please let me know. 

Thanks in advance,
Micha

**Comment by Rensvandeschoot** [comment]
Great you are working on a new feature extractor! What does it do?? (Just curious...)

Does the new model run with any of the pre-loaded benchmark datasets? And, does your data run with the default model?

We also created an issue, see https://github.com/asreview/asreview/issues/1055

**Reply by michagast** [reply]
The feature extractor is part of my masters thesis. I haven't fully implemented it into the extension(as I would not be able to tell which part would be causing what error) but it uses sentiment analysis, named entity regcogntion, presence of certain words an other simpler things to generate a feature matrix/dataframe. I have been able to get decent results with it so far and I haven't even applied the active learning portion yet(the preliminary results are from just the initial training with a small test set).

I have not tried the model with a benchmark dataset but will try to do it asap. The data does run with the default feature_extraction/query/model combo

**Reply by michagast** [reply]
Update: I have tried the feature extraction method with a logistic regression model on the [Van de Schoot](https://www.tandfonline.com/doi/full/10.1080/10705511.2016.1247646) dataset and the same error pops up. Any clue what might be going wrong?

**Comment by PeterLombaers** [comment, marked as answer]
Hi Micha!

Nice to see you're making your own extension! I think there's a couple of things happening. The output of the transform method of the feature extractor should be a matrix (np.array or scipy.sparse.csr_matrix) with one row per text, and the values should all be numeric. That is not what I get as output when I test it. 

Right now, the output of ```sentiment_analysis(text)``` is of the form ```[{'label': 'NEGATIVE', 'score': 0.9959489107131958}]```. So first of all you should make sure that all the features you want to use are numeric (maybe encode 'NEGATIVE' as 0 and 'POSITIVE' as 1?). Next, the function ```np.append``` does not occur in-place (https://numpy.org/doc/stable/reference/generated/numpy.append.html), so when you call ```np.append(result, sentiment_analysis(text))```, result actually stays the same. A fix would be something like ```result = np.append(result, model_output_as_array)```. Even better would be to use ```np.vstack```, since np.append is actually not so efficient to use repeatedly.

Oh, one more tip: It is probably smart to put all code that you only want to execute once in the ```__init__``` method of the class, so that it doesn't get called every time class.transform() is used. So in your case you probably want to move the definition of the variables ```model```, ```tokenizernlp``` and ```sentiment_analysis``` to the ```__init__``` method.

Hope this helps you make some progress! Just make sure to test that the result of the transform method is a matrix with one row for each text and this error should not appear.

**Reply by michagast** [reply]
Thank you! This fixed the issue! And thanks for the general tips, will definitely try to implement them.
